File: tests/cpp/test_multidevice_overlap.cpp
// clang-format off
/*
 * SPDX-FileCopyrightText: Copyright (c) 2025-present NVIDIA CORPORATION & AFFILIATES.
 * All rights reserved.
 * SPDX-License-Identifier: BSD-3-Clause
 */
// clang-format on
#include <gtest/gtest.h>

#include <fusion.h>
#include <ops/all_ops.h>
#include <options.h>
#include <tests/cpp/multidevice.h>

namespace nvfuser {

class StreamTest : public MultiDeviceTest {
 public:
  StreamTest() {
    EnableOptionsGuard::getCurOptions().set(EnableOption::HostIrLowering);
  }
};

// FIXME: make a python test
// Unlike Linear+ReduceScatter, this Linear+Allreduce pattern doesn't benefit
// from ring-based overlapping. However we decompose the compute and the
// communication, a trailing allreduce has to be exposed on the critical path to
// combine the partial results from all participating devices.
TEST_F(StreamTest, RowParallelLinear_Forward) {
  constexpr int64_t h = 2;

  auto fusion = std::make_unique<Fusion>();
  FusionGuard fg(fusion.get());

  TensorView* in = makeContigConcreteTensor({-1, h * 4}, DataType::BFloat16);
  TensorView* w = makeContigConcreteTensor({h, h * 4}, DataType::BFloat16);
  TensorView* out = linear(in, w, nullptr);

  fusion->addInput(in);
  fusion->addInput(w);
  fusion->addOutput(out);

  const int64_t d = communicator_->size();
  const auto mesh = DeviceMesh::createForNumDevices(d);
  for (auto* tv : {in, w}) {
    tv->setDeviceMesh(mesh);
  }

  constexpr int64_t s = 3;
  in->outer_split(0, s);
  in->axis(0)->parallelize(ParallelType::Stream);
  in->outer_split(2, d);
  in->axis(2)->parallelize(ParallelType::DIDx);
  w->outer_split(1, d);
  w->axis(1)->parallelize(ParallelType::DIDx);

  // Fusion IR before segmentation will look like this:
  //
  //   [t, 4h]                                 [h, 4h]
  //   /\  /\                                      /\.
  //  s*  d                                       d
  //                      |
  //                      | linear
  //                      |
  //                          r{4h}
  //                          /  \.
  //                 [t, h, d, r{4h/d}]
  //                 /\.
  //                s
  //                     |
  //                     | sum
  //                     |
  //                  [t, h, r{d}]
  //                  /\.
  //                 s*

  constexpr int64_t t = 6;
  static_assert(t % s == 0);
  at::Tensor in_tensor =
      at::randn({t, h * 4}, tensor_options_.dtype(at::kBFloat16));
  at::Tensor w_tensor =
      at::randn({h, h * 4}, tensor_options_.dtype(at::kBFloat16));
  at::Tensor out_tensor = at::linear(in_tensor, w_tensor);

  at::Tensor sharded_in_tensor = shardTensor(in_tensor, in);
  at::Tensor sharded_w_tensor = shardTensor(w_tensor, w);

  FusionExecutorCache executor_cache(std::move(fusion));
  at::Tensor sharded_out_tensor =
      executor_cache
          .runFusionWithInputs({sharded_in_tensor, sharded_w_tensor})[0]
          .as<at::Tensor>();

  EXPECT_TRUE(at::allclose(sharded_out_tensor, out_tensor));
}

} // namespace nvfuser

File: tests/python/test_multidevice_overlap.py
# SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
# All rights reserved.
# SPDX-License-Identifier: BSD-3-Clause

import pytest
import torch
from enum import Enum, auto
from torch.nn.attention import SDPBackend

import multidevice_fixtures
import nvfuser
import utils
from nvfuser import DataType, FusionDefinition

multidevice_test = multidevice_fixtures.multidevice_test


class OverlapAGMatmulStreamOutermost(FusionDefinition):
    def __init__(self, m, k, n, s, num_devices):
        super().__init__()
        self.m = m
        self.k = k
        self.n = n
        self.s = s
        self._num_devices = num_devices
        self.use_multidevice_executor()

    def definition(self) -> None:
        m, k, n, s, d = (
            self.m,
            self.k,
            self.n,
            self.s,
            self._num_devices,
        )
        self.x = self.define_tensor(
            shape=[s, d, m // (s * d), k], contiguity=True, dtype=DataType.BFloat16
        )
        self.weight = self.define_tensor(
            shape=[n, k], contiguity=True, dtype=DataType.BFloat16
        )
        self.bias = self.define_tensor(
            shape=[n], contiguity=True, dtype=DataType.BFloat16
        )

        self.out = self.ops.linear(
            self.x, self.weight, self.bias
        )  # [s, d, m//(s*d), n]

        self.add_output(self.out)

    def multidevice_schedule(self):
        mesh = nvfuser.DeviceMesh(range(self._num_devices))
        for tv in [
            self.x,
            self.weight,
            self.bias,
            self.out,
        ]:
            self.sched._set_device_mesh(tv, mesh)

        self.sched.parallelize(self.x, 1, nvfuser.ParallelType.mesh_x)
        self.sched.parallelize(self.out, 0, nvfuser.ParallelType.stream)


@pytest.mark.mpi
def test_overlap_allgather_matmul_stream_outermost(multidevice_test, benchmark):
    N_WARMUPS, N_ITERATIONS = 5, 15
    m, k, n, s, d = 1024, 1024, 1024, 8, multidevice_test.size

    torch.cuda.set_device(multidevice_test.local_rank)
    x_unsharded = torch.testing.make_tensor(
        s, d, m // (s * d), k, dtype=torch.bfloat16, device="cpu"
    )
    x = multidevice_test.shard_tensor(
        x_unsharded, 1, nvfuser.DeviceMesh(range(multidevice_test.size))
    )
    weight = torch.testing.make_tensor(n, k, dtype=torch.bfloat16, device="cuda")
    bias = torch.testing.make_tensor(n, dtype=torch.bfloat16, device="cuda")
    ins = [x, weight, bias]
    out_ref = torch.nn.functional.linear(x_unsharded, weight.cpu(), bias.cpu())

    fd = OverlapAGMatmulStreamOutermost(m, k, n, s, d)

    # warmup
    for _ in range(N_WARMUPS):
        outs = fd.execute(ins)
        out = outs[0].local.cpu()
        assert out.dtype == torch.bfloat16
        assert out.shape == torch.Size([s, d, m // (s * d), n])
        torch.testing.assert_close(out, out_ref, rtol=1e-1, atol=1e-1)

    # benchmark
    benchmark.pedantic(lambda: fd.execute(ins), rounds=N_ITERATIONS)
